Documentation/ApiOverview/DependencyInjection/Index.rst

```diff
@@ -12,59 +16,466 @@ Dependency injection
+
+Data object
+  Data objects are the opposite of services. They are *not* available
+  through service containers. Predicate :php:`$container->has()` returns false and they can not be
+  injected. They are instantiated either with :php:`new()` or :php:`GeneralUtility::makeInstance()`.
+  Models are a typical example of data objects.
+
+  *Note:* *Data objects* are *not* "service container aware"
+  and do not support DI. Although the TYPO3 core does not strictly follow this rule in all cases
+  until now the ambition is to get this done over time.
+
+
+.. _supported-ways-of-dependency-injection:
+.. _Using-DI:
+
+Using DI
+========
+
+Now that we have a general understanding of what a service and what a data
+object is, let's turn to usages of services. We will mostly use examples for
+this.
+
+*The general rule is:* Whenever your class has a service dependency to another class,
+one of the following solutions should be used.
+
+
+When to use Dependency Injection in TYPO3
+-----------------------------------------
+
+Class dependencies to services should be injected via constructor injection or
+setter methods. Where possible, Symfony dependency injection should be used for
+all cases where DI is required.
+Non-service "data objects" like Extbase model instances or DTOs should
+be instantiated via :php:`\TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance()`
+if they are non-final and support XCLASSing. For final classes without
+dependencies the `new` keyword can be used.
 
-The recommended way of injecting dependencies is to use constructor injection:
+
+.. _Constructor-injection:
+
+Constructor injection
+---------------------
+
+Assume we're writing a controller that renders a list of users. Those users are
+found using a UserRepository, so the repository service is a direct dependency of the
+controller service. A typical constructor dependency injection to resolve the
+dependency by the framework looks like this:
 
 .. code-block:: php
-   :caption: EXT:some_extension/Classes/SomeClass.php
 
-   public function __construct(Foo $foo, Bar $bar)
+   namespace MyVendor\MyExtension\Controller;
+
+   use MyVendor\MyExtension\Repository\UserRepository;
+
+   class UserController
    {
-       $this->foo = $foo;
-       $this->bar = $bar;
+      protected UserRepository $userRepository;
+
+      public function __construct(UserRepository $userRepository)
+      {
+         $this->userRepository = $userRepository;
+      }
    }
 
-By default all API services shipped by the TYPO3 Core system extensions are available for dependency
-injection.
+((This is straight and easy DI: The controller class is configured to be autowired
+as Symfony DI service container entry (more on that below), Symfony container finds a
+dependency to :php:`UserRepository` when scanning :php:`__construct()` and then knows that
+:php:`UserController` needs an instance of the repository service when the controller
+service is instantiated.)) ← ToDo: Improve understandability
```

Comment:
```suggestion
Here the Symfony container sees a dependency to :php:`UserRepository` when scanning :php:`__construct()`
of the :php:`UserController`. Since autowiring is enabled by default (more on that below), an instance of the
:php:`UserRepository` is created and provided when the controller is created.
```